test/alternator/test_cql_rbac.py

```diff
@@ -0,0 +1,939 @@
+    finally:
+        ret.close()
+
+# A basic test for creating a new role. The ListTables operation is allowed
+# to any role, so it should work in the new role when given the right password
+# and fail with the wrong password.
+def test_new_role(dynamodb, cql):
+    with new_role(cql) as (role, key):
+        with new_dynamodb(dynamodb, role, key) as d:
+            # ListTables should not fail (we don't care what is the result)
+            d.meta.client.list_tables()
+        # Trying to use the wrong key for the new role should fail to perform
+        # any request. The new_dynamodb() function can't detect the error,
+        # it is detected when attempting to perform a request with it.
+        with new_dynamodb(dynamodb, role, 'wrongkey') as d:
+            with pytest.raises(ClientError, match='UnrecognizedClientException'):
+                d.meta.client.list_tables()
+
+# A role without "login" permissions cannot be used to authenticate requests.
+# Reproduces #19735.
+def test_login_false(dynamodb, cql):
+    with new_role(cql, login=False) as (role, key):
+        with new_dynamodb(dynamodb, role, key) as d:
+            with pytest.raises(ClientError, match='UnrecognizedClientException.*login=false'):
+                d.meta.client.list_tables()
+
+# Quote an identifier if it needs to be double-quoted in CQL. Quoting is
+# *not* needed if the identifier matches [a-z][a-z0-9_]*, otherwise it does.
+# double-quotes ('"') in the string are doubled.
+def maybe_quote(identifier):
+    if re.match('^[a-z][a-z0-9_]*$', identifier):
+        return identifier
+    return '"' + identifier.replace('"', '""') + '"'
+
+# Unfortunately, permissions are currently cached for a fixed amount of
+# time, permissions_validity_in_ms, and changing them with GRANT or REVOKE
+# might not have any affect until this timeout. So we need to artificial
+# sleep in temporary_grant() below, and for that we need to know what
+# permissions_validity_in_ms.
+# To make tests fast, it should be configured (e.g. test/cql-pytest/run.py)
+# to be as low as possible, but this test should handle any configured
+# value.
+@cache
+def permissions_validity_in_ms(cql):
+    return 0.001 * int(list(cql.execute("SELECT value FROM system.config WHERE name = 'permissions_validity_in_ms'"))[0][0])
+
+# Convenience context manager for temporarily GRANTing some permission and
+# then revoking it.
+@contextmanager
+def temporary_grant(cql, permission, resource, role):
+    role = maybe_quote(role)
+    cql.execute(f"GRANT {permission} ON {resource} TO {role}")
+    time.sleep(permissions_validity_in_ms(cql)*2)
+    try:
+        yield
+    finally:
+        cql.execute(f"REVOKE {permission} ON {resource} FROM {role}")
+        time.sleep(permissions_validity_in_ms(cql)*2)
+
+@contextmanager
+def temporary_grant_role(cql, role_src, role_dst):
+    role_src = maybe_quote(role_src)
+    role_dst = maybe_quote(role_dst)
+    cql.execute(f"GRANT {role_src} TO {role_dst}")
```

Comment:
I am not sure if this is raft-proof. In general only read barrier guarantees that raft command is applied, but here we have only one node so maybe it is guaranteed (but I doubt). Something to think about or ask someone from raft team.

In other words it may be flaky because sleep doesn't guarantee that GRANT is effective.

Reply:
GRANT is a CQL command, so the user's assumption is that when the execute() returned, it happened. At least on the node where the execution happened (and here we have just one node). If GRANT can return and the user has no idea if it actually happened or not, and if it didn't happen yet, **when** it will happen, it's a problem.
Yes, I know, caching already breaks this feature (GRANT may not apply for permissions_validity_in_ms), but you would like to at least hope that sleeping permissions_validity_in_ms after the GRANT will be enough.

I have to admit that in my tests, sleeping permissions_validity_in_ms was NOT always enough, and I never understood why. I checked that sleeping `permissions_validity_in_ms*2` was enough (I ran the test suite hundreds of times), again I don't know why.

If there's a full-proof way in which I can "GRANT" and verify that it happened that you can suggest, I'll do that.

In test/cql-pytest/test_permission.py (the tests for CQL's RBAC), Piotr Sarna who wrote them took a different approach:  He did a grant, and then called "eventually_authorized" which looped trying some request until eventually it's allowed. In that case, he doesn't care when it's finally allowed (until a ridiculously high timeout of 10 seconds). And similarly for eventually_unauthorized. I didn't like that coding style - I thought it made the test code harder to understand. But that's also an option if we have no limit on how much time GRANT or REVOKE can take to take effect.

CC @kbr-scylla

Reply:
> I am not sure if this is raft-proof. In general only read barrier guarantees that raft command is applied, but here we have only one node so maybe it is guaranteed (but I doubt).

Depends on the implementation. I assume GRANT eventually calls `raft::server::add_entry` underneath with `wait_type::applied` and waits for the returned future to resolve before returning to the user? If so, then the command is guaranteed to have been applied locally. You don't need an additional `read_barrier`.

Reply:
This comment (and the inconclusive reply by @kbr-scylla) caused me to rethink and rewrite the whole way I wrote these tests. I replaced the fixed `permissions_validity_in_ms*2` sleeps by retry loops, that usually finish in around `permissions_validity_in_ms` milliseconds but may retry until a timeout. If we have no strict limit to how much time GRANT may take until it takes affect, we can't do something simpler. 

In the future, I think we should pay more attention to how these APIs are used by real users. Are we telling users "please wait a minute after GRANTing permissions because it may not happen immediately and we can't tell you exactly when it will happen"? Note that schema operations had the same problem and CQL solved this by returning a "schema change" flag in the response and the driver needs to wait until the schema is consistent. Maybe we need to do something similar for RBAC? What did Cassandra do there? How does this work if the role and permissions aren't set in Scylla but in an external LDAP or some such? Anyway, all these questions and concerns are completely orthogonal to **this** PR, which is about Alternator and assumes we are working with the existing GRANT/REVOKE machinery and all its existing features and limitations. So we don't need to fix this now. We just need the test to be correct - and non-flaky - and I believe after my change with retry loops, it is.

Reply:
I'm resolving this conversation - I completely rewrote the retry code following the comment.